extensions/src/persistence/PersistenceCachePlugin.ts

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IPersistence } from "../persistence/IPersistence";
+import { CrossPlatformLock } from "../lock/CrossPlatformLock";
+import { CrossPlatformLockOptions } from "../lock/CrossPlatformLockOptions";
+import { pid } from "process";
+
+/**
+ * MSAL cache plugin which enables callers to write the MSAL cache to disk on Windows,
+ * macOs, and Linux.
+ *
+ * - Persistence can be one of:
+ * - FilePersistence: Writes and reads from an unencrypted file. Can be used on Windows,
+ * macOs, or Linux.
+ * - FilePersistenceWithDataProtection: Used on Windows, writes and reads from file encrypted
+ * with windows dpapi.
+ * - KeychainPersistence: Used on macOs, writes and reads from keychain.
+ * - LibSecretPersistence: Used on linux, writes and reads from secret service API. Requires
+ * libsecret be installed.
+ */
+export class PersistenceCachePlugin {
+
+    public persistence: IPersistence;
+    public lastSync: number;
+    public currentCache: string;
+    public lockFilePath: string;
+
+    private crossPlatformLock: CrossPlatformLock;
+    private readonly lockOptions: CrossPlatformLockOptions;
+
+    constructor(persistence: IPersistence, lockOptions?: CrossPlatformLockOptions) {
+        this.persistence = persistence;
+        this.lockFilePath = `${this.persistence.getFilePath()}.lockfile`;
+        this.lastSync = 0;
+        this.currentCache = null;
+        this.lockOptions = lockOptions;
+    }
+
+    /**
+     * Reads from storage and avoids saves an in memory copy. If persistence has not been updated
+     * since last time data was read, in memory copy is used.
+     */
+    public async readFromStorage(): Promise<string> {
+        console.log("Reading from storage");
+        if (await this.persistence.reloadNecessary(this.lastSync) || this.currentCache == null) {
+            try {
+                console.log("Reload necessary.  Last sync time: " + this.lastSync);
+                this.crossPlatformLock = new CrossPlatformLock(this.lockFilePath, this.lockOptions);
+                await this.crossPlatformLock.lock();
+
+                this.currentCache = await this.persistence.load();
+                this.lastSync = new Date().getTime();
+                console.log("Last sync time updated to: ", this.lastSync);
+            } finally {
+                await this.crossPlatformLock.unlock();
+                delete this.crossPlatformLock;
```

Comment:
I don't think this is the best use of `delete`. Setting to `null` would be better, as it wouldn't actually remove the property.

Reply:
Makes sense. Thinking about it futher, I think the lock object can be reused, so completely removed this.